src/lib/features/webhooklogs.js

```diff
@@ -0,0 +1,181 @@
+import Debug from 'debug';
+import taskcluster from 'taskcluster-client'
+import BulkLog from './bulk_log';
+import getLogsLocationsFromTask from './logs_location.js';
+import slugid from 'slugid'
+import temporary from 'temporary'
+import fs from 'mz/fs'
+
+const debug = Debug('taskcluster-docker-worker:features:webhooklog')
+
+/**
+ * ContinuousReader is a utility class which allows reading a file
+ * which is continuously being written to. A single call to `readUntilEOF`
+ * will read the file until EOF. Subsequent calls will continue reading 
+ * from the point where the previous read ended.
+ */
+class ContinuousReader {
+  constructor(path){
+    this._path = path;
+    this._offset = 0;
+    this._stream = null;
+  }
+
+  readUntilEOF() {
+    // create a new stream that will read until EOF
+    this._stream = fs.createReadStream(this._path, {start: this._offset});
+    // await this promise 
+    return new Promise((resolve, reject) => {
+      let buf = [];
+      this._stream.once('error', reject);
+      // add any new data to the buffer
+      this._stream.on('data', data => buf = buf + data);
+      // once EOF is reached, increment offset, close stream, and resolve
+      // promise with buffer
+      this._stream.on('end', () => {
+        this._offset += this._stream.bytesRead;
+        this._stream.close();
+        return resolve(buf);
+      });
+    });
+  }
+}
+
+/**
+ * Signal is a utility class which is used for notifying readers when data
+ * has been written to the file. It uses a single promise. The constructor
+ * initializes the `_promise` and `_resolve` fields. When `send` is called,
+ * the previous promise is resolved and the fields are reset. When `recv`
+ * is called, the promise is returned. Multiple callers can await the same
+ * promise.
+ */
+class Signal {
+  constructor() {
+    this._resolve = null;
+    this._promise = new Promise(resolve => this._resolve = resolve);
+  }
+
+  send() {
+    this._resolve();
+    this._promise = new Promise(resolve => this._resolve = resolve);
```

Comment:
wrong order... calling `_resolve` could if any of them were synchronous cause a new call to `recv` which would then get the old promise...

You need to swap, create new and then signal..